### PR TITLE
Ports: Add OptiPNG

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
OptiPNG is a neat little tool that optimizes PNG sizes to ridiculous degrees. We like to use it to optimize PNGs before including them in Serenity itself, so it's a nice port to have. OptiPNG is a very cooperative POSIX C program, it compiles and works without any patching on x86_64 and i686 :^)

Here's optipng optimizing one of our ridiculously large `shot` PNGs:

screenshot-2022-06-16-22-34-23

### CODE REVIEW

File: Userland/Libraries/LibC/inttypes.h [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2018-2022, the SerenityOS developers.
```
Comment (timschumi) [inline Userland/Libraries/LibC/inttypes.h:2]: Do we have a history of "merging" Andreas into "the SerenityOS developers"? Or should we just append it to the list instead?

File: Ports/optipng/package.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+
+port="optipng"
+version="0.7.7"
+files="http://prdownloads.sourceforge.net/optipng/optipng-${version}.tar.gz?download optipng-${version}.tar.gz 4f32f233cef870b3f95d3ad6428bfe4224ef34908f1b42b0badf858216654452"
```
Comment (timschumi) [inline Ports/optipng/package.sh:5]: That link seems to be skipping a load balancer (or at least it's not the link that the downloads overview redirects to). `https://downloads.sourceforge.net/optipng/optipng-${version}.tar.gz` seems to work just fine as well.

File: Ports/optipng/package.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+
+port="optipng"
+version="0.7.7"
```
Comment (timschumi) [inline Ports/optipng/package.sh:4]: For consistency, we have recently started to single-quote strings that do not require variable expansion. (Yes, actually documenting and checking that is somewhere on the yakstack.)

File: Ports/optipng/package.sh [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+files="http://prdownloads.sourceforge.net/optipng/optipng-${version}.tar.gz?download optipng-${version}.tar.gz 4f32f233cef870b3f95d3ad6428bfe4224ef34908f1b42b0badf858216654452"
+auth_type=sha256
+useconfigure="true"
```
Comment (timschumi) [inline Ports/optipng/package.sh:7]: Same here.

**Review by timschumi — CHANGES_REQUESTED** [review]
Looking good, just a few small nitpicks! (Testing on my side is yet to be done, but that's a task for tomorrow.)

File: Ports/optipng/package.sh (hunk 2 above)
Comment (kleinesfilmroellchen) [reply Ports/optipng/package.sh:5]: Strange, that's the exact link the website provides

File: Userland/Libraries/LibC/inttypes.h (hunk 1 above)
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibC/inttypes.h:2]: That's actually correct, I thought we had done that but I can't find any instance.

File: Ports/optipng/package.sh (hunk 3 above)
Comment (timschumi) [reply Ports/optipng/package.sh]: GitHub does a bad job at pointing out which lines were targeted in the inline comments, this was also meant to apply for `port` and `auth_type`.

File: Ports/optipng/package.sh [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+
+port="optipng"
```
Comment (linusg) [inline Ports/optipng/package.sh:3]: Single quotes here

File: Ports/optipng/package.sh [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+version='0.7.7'
+files="http://downloads.sourceforge.net/optipng/optipng-${version}.tar.gz optipng-${version}.tar.gz 4f32f233cef870b3f95d3ad6428bfe4224ef34908f1b42b0badf858216654452"
+auth_type=sha256
```
Comment (linusg) [inline Ports/optipng/package.sh:6]: Single quotes here as well

File: Ports/optipng/package.sh [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+useconfigure='true'
+
+configure() {
+	run ./configure
```
Comment (linusg) [inline Ports/optipng/package.sh:10]: 4 spaces for indentation

**Review by linusg — CHANGES_REQUESTED** [review]

File: Ports/optipng/package.sh [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+}
```
Comment (alimpfard) [inline Ports/optipng/package.sh:11]: is this not what the default `configure` does?

Comment (timschumi) [reply Ports/optipng/package.sh:11]: Default configure adds `--host=...`. Maybe we should look into making that optional in the future or move it to `configopts` for consistency.

**Review by timschumi — APPROVED** [review]
LGTM, and runs fine on `i686 GNU`.

**Comment by djwisdom** [comment]
On i686, ./package.sh produced the following error messages:

```sh
ubuntu@PCM-JZBQ2Z2:~/github/sos-pr-checks/Ports/optipng$ ./package.sh
Installing dependencies of optipng...
Fetching optipng...
Downloading URL: http://downloads.sourceforge.net/optipng/optipng-0.7.7.tar.gz
optipng-0.7.7.tar.gz already exists
Expecting sha256sum: 4f32f233cef870b3f95d3ad6428bfe4224ef34908f1b42b0badf858216654452
sha256sum(optipng-0.7.7.tar.gz) = '4f32f233cef870b3f95d3ad6428bfe4224ef34908f1b42b0badf858216654452'
+ tar -xzf optipng-0.7.7.tar.gz (nocd)
+ touch .optipng-0.7.7.tar.gz_extracted
Patching optipng...
Configuring optipng...
+ ./configure
Checking for i686-pc-serenity-gcc...
Using pre-configured libpng...
Configuring zlib...
Building static library libz.a version 1.2.11-optipng with i686-pc-serenity-gcc.
Checking for size_t... Yes.
Checking for off64_t... No.
Checking for fseeko... Yes.
Checking for strerror... Yes.
Checking for unistd.h... Yes.
Checking for stdarg.h... Yes.
Checking whether to use vs[n]printf() or s[n]printf()... using vs[n]printf().
Checking for vsnprintf() in stdio.h... Yes.
Checking for return value of vsnprintf()... Yes.
Checking for attribute(visibility) support... Yes.
Building optipng...
+ make -j6
cd src/optipng && \
make && \
cd ../..
make[1]: Entering directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/optipng'
cd ../zlib && \
make -f Makefile && \
cd ../optipng
cd ../gifread && \
make -f Makefile libgifread.a && \
cd ../optipng
cd ../pnmio && \
make -f Makefile libpnmio.a && \
cd ../optipng
cd ../minitiff && \
make -f Makefile libminitiff.a && \
cd ../optipng
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../cexcept -I../zlib -I../libpng -I../opngreduc -I../pngxtern -o bitset.o bitset.c
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../cexcept -I../zlib -I../libpng -I../opngreduc -I../pngxtern -o ioutil.o ioutil.c
make[2]: Entering directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/gifread'
make[2]: Entering directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/zlib'
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -o gifread.o gifread.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN -I. -c -o example.o test/example.c
make[2]: Entering directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/pnmio'
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -o pnmin.o pnmin.c
make[2]: Entering directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/minitiff'
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../cexcept -I../zlib -I../libpng -I../opngreduc -I../pngxtern -o ratio.o ratio.c
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -o tiffread.o tiffread.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o adler32.o adler32.c
i686-pc-serenity-ar cru libgifread.a gifread.o
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -o pnmout.o pnmout.c
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -o pnmutil.o pnmutil.c
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -o tiffutil.o tiffutil.c
i686-pc-serenity-ranlib libgifread.a
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../cexcept -I../zlib -I../libpng -I../opngreduc -I../pngxtern -o wildargs.o wildargs.c
i686-pc-serenity-ar cru libpnmio.a pnmin.o pnmout.o pnmutil.o
i686-pc-serenity-ar cru libminitiff.a tiffread.o tiffutil.o
make[2]: Leaving directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/gifread'
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o crc32.o crc32.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o deflate.o deflate.c
i686-pc-serenity-ranlib libminitiff.a
i686-pc-serenity-ranlib libpnmio.a
make[2]: Leaving directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/minitiff'
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o infback.o infback.c
make[2]: Leaving directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/pnmio'
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o inffast.o inffast.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o inflate.o inflate.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o inftrees.o inftrees.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o trees.o trees.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o zutil.o zutil.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o compress.o compress.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o uncompr.o uncompr.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o gzclose.o gzclose.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o gzlib.o gzlib.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o gzread.o gzread.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN  -c -o gzwrite.o gzwrite.c
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN -I. -c -o minigzip.o test/minigzip.c
i686-pc-serenity-ar rc libz.a adler32.o crc32.o deflate.o infback.o inffast.o inflate.o inftrees.o trees.o zutil.o compress.o uncompr.o gzclose.o gzlib.o gzread.o gzwrite.o
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN -o example example.o -L. libz.a
i686-pc-serenity-gcc -O3 -DHAVE_HIDDEN -o minigzip minigzip.o -L. libz.a
make[2]: Leaving directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/zlib'
cd ../libpng && \
make -f Makefile PNGLIBCONF_H_PREBUILT=pnglibconf.h.optipng && \
cd ../optipng
make[2]: Entering directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/libpng'
cp pnglibconf.h.optipng pnglibconf.h
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o png.o png.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngerror.o pngerror.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngget.o pngget.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngmem.o pngmem.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngpread.o pngpread.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngread.o pngread.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngrio.o pngrio.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngrtran.o pngrtran.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngrutil.o pngrutil.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngset.o pngset.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngtrans.o pngtrans.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngwio.o pngwio.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngwrite.o pngwrite.c
pngrtran.c:99:1: warning: 'png_rtran_ok' defined but not used [-Wunused-function]
   99 | png_rtran_ok(png_structrp png_ptr, int need_IHDR)
      | ^~~~~~~~~~~~
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngwtran.o pngwtran.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngwutil.o pngwutil.c
i686-pc-serenity-gcc -c -I../zlib  -O2 -Wall -Wextra -o pngtest.o pngtest.c
i686-pc-serenity-ar cru libpng.a png.o pngerror.o pngget.o pngmem.o pngpread.o pngread.o pngrio.o pngrtran.o pngrutil.o pngset.o pngtrans.o pngwio.o pngwrite.o pngwtran.o pngwutil.o
i686-pc-serenity-ranlib libpng.a
i686-pc-serenity-gcc  -L../zlib -o pngtest pngtest.o libpng.a -lz -lm
make[2]: Leaving directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/libpng'
cd ../opngreduc && \
make -f Makefile libopngreduc.a && \
cd ../optipng
cd ../pngxtern && \
make -f Makefile libpngxtern.a && \
cd ../optipng
make[2]: Entering directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/opngreduc'
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../zlib -I../libpng -o opngreduc.o opngreduc.c
make[2]: Entering directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/pngxtern'
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../zlib -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxread.o pngxread.c
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../zlib -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxrbmp.o pngxrbmp.c
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../zlib -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxrgif.o pngxrgif.c
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../zlib -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxrjpg.o pngxrjpg.c
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../zlib -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxrpnm.o pngxrpnm.c
i686-pc-serenity-ar cru libopngreduc.a opngreduc.o
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../zlib -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxrtif.o pngxrtif.c
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../zlib -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxio.o pngxio.c
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../zlib -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxmem.o pngxmem.c
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../zlib -I../libpng -I../gifread -I../pnmio -I../minitiff -o pngxset.o pngxset.c
i686-pc-serenity-ranlib libopngreduc.a
i686-pc-serenity-ar cru libpngxtern.a pngxread.o pngxrbmp.o pngxrgif.o pngxrjpg.o pngxrpnm.o pngxrtif.o pngxio.o pngxmem.o pngxset.o
make[2]: Leaving directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/opngreduc'
i686-pc-serenity-ranlib libpngxtern.a
make[2]: Leaving directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/pngxtern'
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../cexcept -I../zlib -I../libpng -I../opngreduc -I../pngxtern -o optipng.o optipng.c
i686-pc-serenity-gcc -c  -O2 -Wall -Wextra -I../cexcept -I../zlib -I../libpng -I../opngreduc -I../pngxtern -o optim.o optim.c
optim.c: In function 'opng_print_fsize_difference':
optim.c:265:23: error: expected ')' before 'PRIuLEAST64'
  265 |         usr_printf("%" OPNG_FSIZE_PRIu " bytes", difference);
      |                   ~   ^
      |                       )
optim.c: In function 'opng_iterate':
optim.c:1434:51: error: expected ')' before 'PRIuLEAST64'
 1434 |                     usr_printf("\t\tIDAT size = %" OPNG_FSIZE_PRIu "\n",
      |                               ~                   ^
      |                                                   )
optim.c: In function 'opng_finish_iterations':
optim.c:1493:47: error: expected ')' before 'PRIuLEAST64'
 1493 |                 usr_printf("\t\tIDAT size = %" OPNG_FSIZE_PRIu,
      |                           ~                   ^
      |                                               )
optim.c: In function 'opng_optimize_impl':
optim.c:1687:41: error: expected ')' before 'PRIuLEAST64'
 1687 |         usr_printf("Input IDAT size = %" OPNG_FSIZE_PRIu " bytes\n",
      |                   ~                     ^
      |                                         )
optim.c:1689:37: error: expected ')' before 'PRIuLEAST64'
 1689 |     usr_printf("Input file size = %" OPNG_FSIZE_PRIu " bytes\n",
      |               ~                     ^
      |                                     )
optim.c:1819:40: error: expected ')' before 'PRIuLEAST64'
 1819 |     usr_printf("\nOutput IDAT size = %" OPNG_FSIZE_PRIu " bytes",
      |               ~                        ^
      |                                        )
optim.c:1828:40: error: expected ')' before 'PRIuLEAST64'
 1828 |     usr_printf("\nOutput file size = %" OPNG_FSIZE_PRIu " bytes (",
      |               ~                        ^
      |                                        )
make[1]: *** [Makefile:103: optim.o] Error 1
make[1]: Leaving directory '/home/ubuntu/github/sos-pr-checks/Ports/optipng/optipng-0.7.7/src/optipng'
make: *** [Makefile:2: all] Error 2
```

**Comment by timschumi** [comment]
Your Serenity installation is outdated, you need to build and install again.

**Comment by EWouters** [comment]
It fails to build on Mac unfortunately:

```
x86_64-pc-serenity-gcc -O3 -DHAVE_HIDDEN -I. -c -o minigzip.o test/minigzip.c
libtool -o libz.a adler32.o crc32.o deflate.o infback.o inffast.o inflate.o inftrees.o trees.o zutil.o compress.o uncompr.o gzclose.o gzlib.o gzread.o gzwrite.o
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: adler32.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: crc32.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: deflate.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: infback.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: inffast.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: inflate.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: inftrees.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: trees.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: zutil.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: compress.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: uncompr.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: gzclose.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: gzlib.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: gzread.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: gzwrite.o is not an object file (not allowed in a library)
make[2]: *** [libz.a] Error 1
make[1]: *** [../zlib/libz.a] Error 2
make: *** [all] Error 2
```

**Comment by BertalanD** [comment]
Why is it using host `ar`. Do we not set the `AR` environment variable?

**Pull request merged**